### PR TITLE
ANNOT: Extend mutable check annotator and add quick fix

### PR DESCRIPTION
**Pull request by farodin91**
This checks field, functions and assignment and allow quick fix.

### CODE REVIEW

**Comment by farodin91** [comment]
@matklad Do you like to review my new PR?

**Comment by matklad** [comment]
>Do you like to review my new PR?

Sure, but this most likely will happen tomorrow! I'm a bit overloaded at the moment :(

**Comment by farodin91** [comment]
No problem!

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        updateMutable(project, binding, mutable)
+    }
+}
+
+fun updateMutable(project: Project, binding: RsNamedElement, mutable: Boolean = true) {
+    when (binding) {
+        is RsPatBinding -> {
+            val parameter = binding.parentOfType<RsValueParameter>()
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt]: Judging by the test it should work correctly. Let's perhaps add a couple of tests with complex patterns in functions, to make sure we don't do something funny? 

Possible tests would be

```Rust
fn foo((x, y): (i32, i32)) {
    x = 92;
}

fn foo(&mut (ref x, ): &mut (String, )) {
    x.push_str("x");
}
```

Probably there are other interesting ones!

**Comment by matklad** [comment]
Looks good! I am not entirely sure we correctly handle all possible permutations of `ref mut &` in patterns in the fix.

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt (hunk 1 above)
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt]: One case need to be updated. Only Pattern update doesn't work currently.

**Comment by farodin91** [comment]
Todo:
- Support for reassign annotator in a if else clause. Correct code:
```rust
fn main() {
  let a;
  if(true) {
    a = 10;
  } else {
    a = 20;
  }
}
```
- Support for quick fix in ValueParameter for pattern part
```rust
struct S;
impl S {
  fn foo(&mut self) {
  }
}
fn test((x, y): (&S, &mut S)) {
 x.foo();
}
```

**Comment by farodin91** [comment]
Resolved all test problems.

**Comment by farodin91** [comment]
@mkaput r+

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -412,6 +432,33 @@ class RsErrorAnnotator : Annotator, HighlightRangeExtension {
...
+                holder.createErrorAnnotation(
+                    expr, message
+                ).registerFix(AddMutableFix(declaration))
+            }
+            else -> {
+                holder.createErrorAnnotation(
+                    expr, message
+                )
+            }
+        }
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: These lines look too copy-paste'y for me. Why not:

```kotlin
val annotation = holder.createErrorAnnotation(expr, message)
val declaration = expr.declaration()
if (declaration is RsSelfParameter || declaration is RsPathBinding) {
    annotation.registerFix(AddMutableFix(declaration)
}
```

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 3]
```diff
@@ -520,17 +567,66 @@ private fun RsCallExpr.expectedParamsCount(): Int? {
...
+
+private fun RsExpr.declaration(): RsCompositeElement? {
+    if (this is RsPathExpr) {
+        return this.path.reference.resolve()
+    }
+    if (this is RsFieldExpr) {
+        return this.expr.declaration()
+    }
+    return null
+}
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: I'm not sure, but shouldn't these 2 functions reside in `ResolveEngine`? What's your opinion @matklad ?

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 4]
```diff
@@ -520,17 +567,66 @@ private fun RsCallExpr.expectedParamsCount(): Int? {
...
+            false
+        }
+        is RsFieldExpr -> (this.expr.type as? RustReferenceType)?.mutable ?: true
+        is RsUnaryExpr -> {
+            if (this.mul != null) {
+                return true
+            } else {
+                return this.expr?.isMutable() ?: true
+            }
+        }
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]:
```kotlin
is RsUnaryExpr -> mul != null || (expr != null && expr.isMutable())
```

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsNamedElement
+import org.rust.lang.core.psi.ext.parentOfType
+
+class AddMutableFix(val binding: RsNamedElement) : LocalQuickFixAndIntentionActionOnPsiElement(binding) {
+    override fun getFamilyName(): String = text
+    override fun getText(): String = "Add mutable"
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddMutableFix.kt]: (Nitpicking) Personally I don't like naming this quickfix *Add mutable*. Maybe *Make variable mutable* or sth?

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt (hunk 3 above)
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: Also maybe 
```kotlin
private fun RsExpr.declaration(): RsCompositeElement? = when(this) ...
```

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 6]
```diff
@@ -520,17 +567,66 @@ private fun RsCallExpr.expectedParamsCount(): Int? {
...
+    if (this is RsFieldExpr) {
+        return this.expr.declaration()
+    }
+    return null
+}
+
 private fun RsExpr.isMutable(): Boolean {
-    val path = (this as? RsPathExpr)?.path ?: return true
-    val declaration = path.reference.resolve()
+    return when (this) {
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]:
```kotlin
private fun RsExpr.isMutable(): Boolean = when(this) ...
```

File: src/test/kotlin/org/rust/ide/annotator/RsExpressionAnnotatorTest.kt [hunk 7]
```diff
@@ -150,8 +150,257 @@ class RsExpressionAnnotatorTest : RsAnnotatorTestBase() {
...
+        }
+    """)
+
+    fun `test let some from mutable reference`() = checkErrors("""
+        fn foo(optional: Option<&mut Vec<String>>) {
+            if let Some(x) = optional {
+                *x = "str".to_string();
+            }
+        }
+    """)
```
Comment (mkaput) [inline src/test/kotlin/org/rust/ide/annotator/RsExpressionAnnotatorTest.kt]: :+1: I like these tests 😄

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt (hunk 6 above)
Comment (farodin91) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: how to replace return in a expression.

Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: Oh, sorry. I forgot about them 😅

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 8]
```diff
@@ -520,17 +554,56 @@ private fun RsCallExpr.expectedParamsCount(): Int? {
...
+            if (letExpr != null && letExpr.eq == null) {
+                return true
+            }
+            if (type is RustUnknownType) {
+                return true
+            }
+            false
+        }
+        is RsFieldExpr -> (this.expr.type as? RustReferenceType)?.mutable ?: true
+        is RsUnaryExpr -> mul != null || (expr != null && expr?.isMutable()!!)
```
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt]: Are these `!!` necessary to compile? If so, then you can stick with older version `(expr?.isMutable() ?: true)`

File: src/test/kotlin/org/rust/ide/annotator/RsExpressionAnnotatorTest.kt [hunk 9]
```diff
@@ -150,8 +150,257 @@ class RsExpressionAnnotatorTest : RsAnnotatorTestBase() {
...
+            }
+            a = 5;//FIXME:  this line should fail
+        }
+    """)
+
+    fun `test mutable used at reassign`() = checkErrors("""
+        fn main() {
+            let mut x;
+            x = 3;
+            x = 5;//FIXME: this line should fail
```
Comment (alygin) [inline src/test/kotlin/org/rust/ide/annotator/RsExpressionAnnotatorTest.kt]: Why should this one fail? `x` is mutable, we're free to reassign it.

**Comment by matklad** [comment]
Looks great @farodin91! Could you rebase this branch to fix the conflicts with `unsafe` errors?

**Comment by farodin91** [comment]
@matklad Rebased!

**Comment by matklad** [comment]
Thanks!

**Pull request merged**